test_common/harness/kernelHelpers.cpp

```diff
@@ -809,6 +811,56 @@ int create_single_kernel_helper(cl_context context,
                                 const char *buildOptions,
                                 const bool openclCXX)
 {
+
+    // Check the build options for the -cl-std option.
+    if (!buildOptions || !strstr(buildOptions, "-cl-std"))
+    {
+        // If the build option isn't present add it using the latetest OpenCL-C
+        // version supported by the device. This allows calling code to force a
+        // particular CL C version if it is required, but also means that
+        // callers need not specify a version if they want to assume the most
+        // recent CL C.
+
+        // First get all the devices in the context and find the minumum
+        // supported OpenCL C version.
+        size_t devices_size_in_bytes{};
+        int error = clGetContextInfo(context, CL_CONTEXT_DEVICES, 0, nullptr,
+                                     &devices_size_in_bytes);
+        test_error(error, "clGetDeviceInfo failed for CL_CONTEXT_DEVICES");
+        std::vector<cl_device_id> devices(devices_size_in_bytes
+                                          / sizeof(cl_device_id));
+        error =
+            clGetContextInfo(context, CL_CONTEXT_DEVICES, devices_size_in_bytes,
+                             devices.data(), nullptr);
+        Version min_version{ (std::numeric_limits<int>::max)(),
+                             (std::numeric_limits<int>::max)() };
+        std::for_each(devices.begin(), devices.end(),
+                      [&min_version](cl_device_id device) {
+                          auto version = get_device_cl_version(device);
+                          min_version = (std::min)(version, min_version);
```

Comment:
If there are two devices, one supporting OpenCL 3.0 and one supporting OpenCL 2.0, then -cl-std=CL2.0 will be chosen.

The 3.0 device might not necessarily support OpenCL C 2.0, meaning that we'd need to drop to OpenCL C 1.2.

Reply:
Nice catch, I completely overlooked that. I've updated the PR to use 1.2 in the case that it finds a 2.X device and there is already a 3.0 device, or it finds a 3.0 device and there is already a 2.X device. The logic looks a bit convoluted now but I'm not sure if there is a cleaner way.

Reply:
Thanks, looks good. I don't think there is much that could be done to make it cleaner, there is not getting around just having to loop through every device and compare against what has already been discovered.